plugin/backend.go

```diff
@@ -58,94 +65,167 @@ func Backend() *GcpAuthBackend {
-		creds, err = google.FindDefaultCredentials(ctx, b.oauthScopes...)
+	return client.(*iam.Service), nil
+}
+
+// ComputeClient returns a new Compute client. The client is cached.
+func (b *GcpAuthBackend) ComputeClient(s logical.Storage) (*compute.Service, error) {
+	httpClient, err := b.httpClient(s)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to create Compute HTTP client: {{err}}", err)
+	}
+
+	client, err := b.cache.Fetch("compute", cacheTime, func() (interface{}, error) {
+		client, err := compute.New(httpClient)
 		if err != nil {
-			return nil, errwrap.Wrapf(
-				"credentials were not configured and Vault could not find "+
-					"Application Default Credentials (ADC). Either set ADC or "+
-					"configure this auth backend at auth/$MOUNT/config "+
-					"(default auth/gcp/config). Error: {{err}}", err)
+			return nil, errwrap.Wrapf("failed to create Compute client: {{err}}", err)
 		}
+		client.UserAgent = useragent.String()
+
+		return client, nil
+	})
+	if err != nil {
+		return nil, err
 	}
 
-	cleanCtx := context.WithValue(ctx, oauth2.HTTPClient, cleanhttp.DefaultClient())
-	client := oauth2.NewClient(cleanCtx, creds.TokenSource)
-	return client, nil
+	return client.(*compute.Service), nil
 }
 
-func (b *GcpAuthBackend) newGcpClients(ctx context.Context, s logical.Storage) (*clientHandles, error) {
-	httpC, err := b.httpClient(ctx, s)
+// CRMClient returns a new Cloud Resource Manager client. The client is cached.
+func (b *GcpAuthBackend) CRMClient(s logical.Storage) (*cloudresourcemanager.Service, error) {
+	httpClient, err := b.httpClient(s)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to create Cloud Resource Manager HTTP client: {{err}}", err)
+	}
+
+	client, err := b.cache.Fetch("crm", cacheTime, func() (interface{}, error) {
+		client, err := cloudresourcemanager.New(httpClient)
+		if err != nil {
+			return nil, errwrap.Wrapf("failed to create Cloud Resource Manager client: {{err}}", err)
+		}
+		client.UserAgent = useragent.String()
+
+		return client, nil
+	})
 	if err != nil {
-		return nil, errwrap.Wrapf("could not obtain HTTP client: {{err}}", err)
+		return nil, err
 	}
 
-	iamClient, err := iam.New(httpC)
+	return client.(*cloudresourcemanager.Service), nil
+}
+
+// httpClient returns a new http.Client that is authenticated using the provided
+// credentials. The underlying httpClient is cached among all clients.
+func (b *GcpAuthBackend) httpClient(s logical.Storage) (*http.Client, error) {
+	creds, err := b.credentials(s)
 	if err != nil {
-		return nil, fmt.Errorf(clientErrorTemplate, "IAM", err)
+		return nil, errwrap.Wrapf("failed to create oauth2 http client: {{err}}", err)
 	}
-	iamClient.UserAgent = useragent.String()
 
-	gceClient, err := compute.New(httpC)
+	client, err := b.cache.Fetch("HTTPClient", cacheTime, func() (interface{}, error) {
+		b.Logger().Debug("creating oauth2 http client")
+		ctx := context.WithValue(context.Background(), oauth2.HTTPClient, cleanhttp.DefaultClient())
```

Comment:
Since we're redoing this, perhaps we should incorporate`context.WithCancel()`, save the `CancelFunc` on backend object, and implement `Clean()` to invoke it if set.

Reply:
I had that originally before I built the `cache` package, but I think it'd be really complicated with the race conditions. `Clean()` would have to acquire a lock on the HTTP client key before calling cancel, and I'm not sure it's worth the complexity given how quickly most of these requests will complete?

Reply:
It can be left as is.